File: tests/test_http.py
"""Tests for HTTP-layer functions (search, fetch_info, download_torrent).

Uses the `responses` library to intercept requests.Session calls so no
real network traffic is made.
"""

from __future__ import annotations

from pathlib import Path

import pytest
import responses as rsps_lib

import io

from iptorrents.download import download_torrent, stream_torrent
from iptorrents.info import fetch_info
from iptorrents.search import search
from iptorrents.session import BASE_URL, make_session

from .fixtures import (
    INFO_HTML_MOVIE,
    INFO_HTML_SOFTWARE,
    SEARCH_HTML,
    SEARCH_HTML_LOGGED_OUT,
    SESSION_OK_HTML,
)


@pytest.fixture()
def session():
    """Unauthenticated-but-valid session (cookies are fake — responses lib intercepts)."""
    return make_session(uid="fake_uid", pass_cookie="fake_pass", cf_clearance="fake_cf")


# ---------------------------------------------------------------------------
# search()
# ---------------------------------------------------------------------------


class TestSearch:
    @rsps_lib.activate
    def test_returns_results(self, session):
        rsps_lib.add(rsps_lib.GET, f"{BASE_URL}/t", body=SEARCH_HTML, status=200)
        results = search(session, "blade runner")
        assert len(results) == 2

    @rsps_lib.activate
    def test_movie_result_fields(self, session):
        rsps_lib.add(rsps_lib.GET, f"{BASE_URL}/t", body=SEARCH_HTML, status=200)
        results = search(session, "blade runner")
        movie = results[0]
        assert movie.id == 111222
        assert movie.category == "Movie/HD"
        assert movie.seeders == 987

    @rsps_lib.activate
    def test_limit_is_forwarded(self, session):
        rsps_lib.add(rsps_lib.GET, f"{BASE_URL}/t", body=SEARCH_HTML, status=200)
        results = search(session, "blade runner", limit=1)
        assert len(results) == 1

    @rsps_lib.activate
    def test_sort_param_sent(self, session):
        rsps_lib.add(rsps_lib.GET, f"{BASE_URL}/t", body=SEARCH_HTML, status=200)
        search(session, "ubuntu", sort="seeders")
        assert len(rsps_lib.calls) == 1
        assert "o=seeders" in rsps_lib.calls[0].request.url

    @rsps_lib.activate
    def test_network_error_exits(self, session):
        rsps_lib.add(rsps_lib.GET, f"{BASE_URL}/t", body=ConnectionError("offline"))
        with pytest.raises(SystemExit):
            search(session, "anything")

    @rsps_lib.activate
    def test_logged_out_page_exits(self, session):
        rsps_lib.add(rsps_lib.GET, f"{BASE_URL}/t", body=SEARCH_HTML_LOGGED_OUT, status=200)
        with pytest.raises(SystemExit):
            search(session, "anything")


# ---------------------------------------------------------------------------
# fetch_info()
# ---------------------------------------------------------------------------


class TestFetchInfo:
    @rsps_lib.activate
    def test_movie_parsed_correctly(self, session):
        rsps_lib.add(rsps_lib.GET, f"{BASE_URL}/t/111222", body=INFO_HTML_MOVIE, status=200)
        info = fetch_info(session, 111222)
        assert info.id == 111222
        assert "Blade Runner" in info.name
        assert info.seeders == 987
        assert info.genre == ("Sci-Fi", "Drama")
        assert "imdb.com" in info.imdb_url

    @rsps_lib.activate
    def test_software_no_crash(self, session):
        """fetch_info must not crash for torrents without movie metadata."""
        rsps_lib.add(rsps_lib.GET, f"{BASE_URL}/t/333444", body=INFO_HTML_SOFTWARE, status=200)
        info = fetch_info(session, 333444)
        assert info.id == 333444
        assert info.genre == ()
        assert info.plot == ""
        assert info.imdb_url == ""

    @rsps_lib.activate
    def test_http_error_exits(self, session):
        rsps_lib.add(rsps_lib.GET, f"{BASE_URL}/t/99999", status=403)
        with pytest.raises(SystemExit):
            fetch_info(session, 99999)

    @rsps_lib.activate
    def test_network_error_exits(self, session):
        rsps_lib.add(rsps_lib.GET, f"{BASE_URL}/t/99999", body=ConnectionError("offline"))
        with pytest.raises(SystemExit):
            fetch_info(session, 99999)


# ---------------------------------------------------------------------------
# download_torrent()
# ---------------------------------------------------------------------------


FAKE_TORRENT_BYTES = b"d8:announce35:http://tracker.example.com/announcee"


class TestStreamTorrent:
    @rsps_lib.activate
    def test_writes_bytes_to_stream(self, session):
        dl_url = f"{BASE_URL}/download.php/111222/file.torrent"
        rsps_lib.add(rsps_lib.GET, dl_url, body=FAKE_TORRENT_BYTES, status=200)
        buf = io.BytesIO()
        stream_torrent(session, dl_url, buf)
        assert buf.getvalue() == FAKE_TORRENT_BYTES

    @rsps_lib.activate
    def test_nothing_written_on_http_error(self, session):
        dl_url = f"{BASE_URL}/download.php/0/bad.torrent"
        rsps_lib.add(rsps_lib.GET, dl_url, status=403)
        buf = io.BytesIO()
        with pytest.raises(SystemExit):
            stream_torrent(session, dl_url, buf)
        assert buf.getvalue() == b""

    @rsps_lib.activate
    def test_no_file_created_on_disk(self, session, tmp_path):
        dl_url = f"{BASE_URL}/download.php/111222/file.torrent"
        rsps_lib.add(rsps_lib.GET, dl_url, body=FAKE_TORRENT_BYTES, status=200)
        buf = io.BytesIO()
        stream_torrent(session, dl_url, buf)
        assert list(tmp_path.iterdir()) == []


class TestDownloadTorrent:
    @rsps_lib.activate
    def test_saves_file_to_disk(self, session, tmp_path):
        dl_url = f"{BASE_URL}/download.php/111222/Blade.Runner.2049.torrent"
        rsps_lib.add(
            rsps_lib.GET,
            dl_url,
            body=FAKE_TORRENT_BYTES,
            status=200,
            headers={"Content-Disposition": 'attachment; filename="Blade.Runner.2049.torrent"'},
            content_type="application/x-bittorrent",
        )
        out = download_torrent(session, dl_url, dest_dir=tmp_path)
        assert out.exists()
        assert out.read_bytes() == FAKE_TORRENT_BYTES
        assert out.suffix == ".torrent"

    @rsps_lib.activate
    def test_filename_from_content_disposition(self, session, tmp_path):
        dl_url = f"{BASE_URL}/download.php/111222/file.torrent"
        rsps_lib.add(
            rsps_lib.GET,
            dl_url,
            body=FAKE_TORRENT_BYTES,
            status=200,
            headers={"Content-Disposition": 'attachment; filename="custom-name.torrent"'},
        )
        out = download_torrent(session, dl_url, dest_dir=tmp_path)
        assert out.name == "custom-name.torrent"

    @rsps_lib.activate
    def test_filename_falls_back_to_url(self, session, tmp_path):
        dl_url = f"{BASE_URL}/download.php/111222/fallback-name.torrent"
        rsps_lib.add(rsps_lib.GET, dl_url, body=FAKE_TORRENT_BYTES, status=200)
        out = download_torrent(session, dl_url, dest_dir=tmp_path)
        assert out.name == "fallback-name.torrent"

    @rsps_lib.activate
    def test_explicit_filename_override(self, session, tmp_path):
        dl_url = f"{BASE_URL}/download.php/111222/whatever.torrent"
        rsps_lib.add(rsps_lib.GET, dl_url, body=FAKE_TORRENT_BYTES, status=200)
        out = download_torrent(session, dl_url, dest_dir=tmp_path, filename="my-override.torrent")
        assert out.name == "my-override.torrent"

    @rsps_lib.activate
    def test_torrent_extension_appended_if_missing(self, session, tmp_path):
        dl_url = f"{BASE_URL}/download.php/111222/noext"
        rsps_lib.add(
            rsps_lib.GET,
            dl_url,
            body=FAKE_TORRENT_BYTES,
            status=200,
        )
        out = download_torrent(session, dl_url, dest_dir=tmp_path, filename="noext")
        assert out.name == "noext.torrent"

    @rsps_lib.activate
    def test_http_error_exits(self, session, tmp_path):
        dl_url = f"{BASE_URL}/download.php/0/bad.torrent"
        rsps_lib.add(rsps_lib.GET, dl_url, status=404)
        with pytest.raises(SystemExit):
            download_torrent(session, dl_url, dest_dir=tmp_path)

    @rsps_lib.activate
    def test_dest_dir_created_if_missing(self, session, tmp_path):
        dl_url = f"{BASE_URL}/download.php/111222/file.torrent"
        rsps_lib.add(rsps_lib.GET, dl_url, body=FAKE_TORRENT_BYTES, status=200)
        new_dir = tmp_path / "deep" / "nested"
        assert not new_dir.exists()
        out = download_torrent(session, dl_url, dest_dir=new_dir)
        assert new_dir.exists()
        assert out.exists()

File: tests/test_http.rs
mod support;

use std::io::Cursor;

use httpmock::Method::GET;
use httpmock::MockServer;
use iptorrents_cli::active::active;
use iptorrents_cli::download::{download_torrent, stream_torrent};
use iptorrents_cli::http::{IptClient, verify_session};
use iptorrents_cli::info::fetch_info;
use iptorrents_cli::models::AuthConfig;
use iptorrents_cli::search::search;
use tempfile::tempdir;

use crate::support::fixtures::{INFO_HTML_MOVIE, PEERS_HTML, SEARCH_HTML, SEARCH_HTML_LOGGED_OUT};

const FAKE_TORRENT_BYTES: &[u8] = b"d8:announce35:http://tracker.example.com/announcee";

fn test_client(base_url: String) -> IptClient {
    IptClient::new(
        AuthConfig {
            uid: "fake_uid".to_string(),
            pass_cookie: "fake_pass".to_string(),
            cf_clearance: Some("fake_cf".to_string()),
        },
        base_url,
    )
    .expect("client should build")
}

#[test]
fn verify_session_ok() {
    let server = MockServer::start();
    let _m = server.mock(|when, then| {
        when.method(GET).path("/t");
        then.status(200).body("ok");
    });

    let client = test_client(server.base_url());
    verify_session(&client).expect("session should verify");
}

#[test]
fn verify_session_rejects_sign_in_page() {
    let server = MockServer::start();
    let _m = server.mock(|when, then| {
        when.method(GET).path("/t");
        then.status(200)
            .body("<html><body>Please sign in to continue</body></html>");
    });

    let client = test_client(server.base_url());
    let err = verify_session(&client).unwrap_err();
    assert!(err.to_string().contains("invalid or expired"));
}

#[test]
fn search_returns_results() {
    let server = MockServer::start();

    let _verify = server.mock(|when, then| {
        when.method(GET).path("/t");
        then.status(200).body(SEARCH_HTML);
    });

    let client = test_client(server.base_url());
    let results = search(&client, "blade runner", 25, None).expect("search should succeed");
    assert_eq!(results.len(), 3);
}

#[test]
fn search_sends_sort_param() {
    let server = MockServer::start();

    let sorted = server.mock(|when, then| {
        when.method(GET)
            .path("/t")
            .query_param("q", "ubuntu")
            .query_param("o", "seeders");
        then.status(200).body(SEARCH_HTML);
    });

    let client = test_client(server.base_url());
    let _ = search(&client, "ubuntu", 25, Some("seeders")).expect("search should succeed");
    sorted.assert();
}

#[test]
fn search_logged_out_errors() {
    let server = MockServer::start();
    let _m = server.mock(|when, then| {
        when.method(GET).path("/t");
        then.status(200).body(SEARCH_HTML_LOGGED_OUT);
    });

    let client = test_client(server.base_url());
    let err = search(&client, "anything", 25, None).unwrap_err();
    assert!(err.to_string().contains("not logged in"));
}

#[test]
fn fetch_info_parses_movie() {
    let server = MockServer::start();
    let _m = server.mock(|when, then| {
        when.method(GET).path("/t/111222");
        then.status(200).body(INFO_HTML_MOVIE);
    });

    let client = test_client(server.base_url());
    let info = fetch_info(&client, 111_222).expect("info should parse");
    assert_eq!(info.id, 111_222);
    assert!(info.name.contains("Blade Runner"));
    assert_eq!(info.seeders, 987);
}

#[test]
fn stream_torrent_writes_bytes() {
    let server = MockServer::start();
    let download_path = "/download.php/111222/file.torrent";
    let _m = server.mock(|when, then| {
        when.method(GET).path(download_path);
        then.status(200)
            .header("content-type", "application/x-bittorrent")
            .body(FAKE_TORRENT_BYTES);
    });

    let client = test_client(server.base_url());
    let mut cursor = Cursor::new(Vec::new());
    stream_torrent(
        &client,
        &format!("{}{}", server.base_url(), download_path),
        &mut cursor,
    )
    .expect("stream should work");
    assert_eq!(cursor.into_inner(), FAKE_TORRENT_BYTES);
}

#[test]
fn download_torrent_saves_file() {
    let server = MockServer::start();
    let download_path = "/download.php/111222/file.torrent";
    let _m = server.mock(|when, then| {
        when.method(GET).path(download_path);
        then.status(200)
            .header(
                "content-disposition",
                "attachment; filename=\"custom-name.torrent\"",
            )
            .body(FAKE_TORRENT_BYTES);
    });

    let client = test_client(server.base_url());
    let tmp = tempdir().expect("temp dir");
    let out = download_torrent(
        &client,
        &format!("{}{}", server.base_url(), download_path),
        tmp.path(),
        None,
    )
    .expect("download should work");

    assert!(out.exists());
    assert_eq!(
        out.file_name().and_then(|n| n.to_str()),
        Some("custom-name.torrent")
    );
    assert_eq!(std::fs::read(out).expect("read output"), FAKE_TORRENT_BYTES);
}

#[test]
fn active_returns_seeding_and_leeching() {
    let server = MockServer::start();

    let _m = server.mock(|when, then| {
        when.method(GET).path("/peers");
        then.status(200).body(PEERS_HTML);
    });

    let client = test_client(server.base_url());
    let results = active(&client).expect("active should succeed");

    assert_eq!(results.seeding.len(), 2);
    assert_eq!(results.leeching.len(), 1);
}
